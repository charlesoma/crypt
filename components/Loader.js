import React, { Component } from 'react'
import { View, StyleSheet, ActivityIndicator } from 'react-native'

const Loader = () => {
    return (
        <View style={styles.overlay}>
            <View style={styles.loaderContainer}>
                <ActivityIndicator size="large" color="#fff" />
            </View>
        </View>
    )
}

export default Loader;

const styles = StyleSheet.create({
    overlay: {
        height: '100%',
        width: '100%',
        display: 'flex',
        alignItems: 'center',
        zIndex: 5,
        elevation: 5
    },
    loaderContainer: {
        height: 80,
        width: 80,
        backgroundColor: '#1B2845',
        display: 'flex',
        justifyContent: 'center',
        borderRadius: 12,
        position: 'absolute',
        top: '35%',
    }
})
