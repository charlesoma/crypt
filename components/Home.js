import React from 'react';
import { StyleSheet, Text, View, ScrollView, TouchableOpacity, RefreshControl } from 'react-native';
import { Header } from 'react-native-elements';
import { connect } from "react-redux";
import { fetchCrypto, fetchMoreCrypto } from "../actions/home";
import moment from 'moment';
import Loader from "../components/Loader";

export class Home extends React.Component {
    state = {
        data: [],
        start: 1,
        clicked: false,
        refreshing: false,
    }

    async componentDidMount() {
        await this.props.fetchCrypto()
        this.setState({
            data: this.props.data,
            start: 1,
            clicked: false
        })
    }

    more = async () => {
        this.setState({
            clicked: true
        })
        await this.props.fetchMoreCrypto(this.state.start + 10)
        this.setState({
            more: this.props.activeMore,
            start: this.state.start + 10,
            data: this.state.data.concat(this.props.dataMore),
            clicked: false
        })
    }

    _onRefresh = async () => {
        this.setState({ refreshing: true });
        await this.componentDidMount().then(() => {
            this.setState({ refreshing: false });
        });
    }

    reload = () => {
        this.componentDidMount();
        this.setState({
            clicked: true
        })
    }

    render() {
        const { data } = this.state;
        return (
            <View>
                <Header
                    containerStyle={styles.header}
                    centerComponent={{ text: 'Crypt', style: { color: '#fff', fontSize: 24 } }}
                />
                {
                    this.props.data === null ?
                        <Loader />
                        :
                        <View>
                            {this.props.error ?
                                <View style={styles.error}>
                                    <Text style={styles.errorText}>{this.props.data[0]}</Text>
                                    <TouchableOpacity
                                        onPress={this.reload}
                                        style={styles.btn}
                                    >
                                        {
                                            this.state.clicked ?
                                                <Text style={styles.btnText}>Loading...</Text>
                                                :
                                                <Text style={styles.btnText}>Reload</Text>
                                        }
                                    </TouchableOpacity>
                                </View>
                                :
                                <ScrollView
                                    contentContainerStyle={styles.body}
                                    refreshControl={
                                        <RefreshControl
                                            refreshing={this.state.refreshing}
                                            onRefresh={this._onRefresh}
                                        />
                                    }
                                >
                                    {data.map((data, index) =>
                                        <View style={styles.card} key={index}>
                                            <Text style={styles.title}>{data.symbol}</Text>
                                            <View style={styles.divider}></View>
                                            <View style={styles.cardTop}>
                                                <View style={styles.row}>
                                                    <Text style={styles.label}>Cryptocurrency:</Text>
                                                    <Text style={styles.value}> {data.name}</Text>
                                                </View>
                                                <View style={styles.row}>
                                                    <Text style={styles.label}>Symbol:</Text>
                                                    <Text style={styles.value}> {data.symbol}</Text>
                                                </View>
                                            </View>
                                            <View style={styles.rowSingle}>
                                                <Text style={styles.label}>Last updated:</Text>
                                                <Text style={styles.value}> {moment(data.quote.USD.last_updated).format('MMM Do YYYY, h:mm:ss a')}</Text>
                                            </View>
                                            <View style={styles.rowSingle}>
                                                <Text style={styles.label}>Price:</Text>
                                                <Text style={styles.value}> ${Math.round(data.quote.USD.price * 100) / 100}</Text>
                                            </View>
                                            <View style={styles.cardBody}>
                                                <View style={styles.row}>
                                                    <Text style={styles.label}>% change in 1h:</Text>
                                                    <Text style={styles.value}> {Math.round(data.quote.USD.percent_change_1h * 100) / 100}</Text>
                                                </View>
                                                <View style={styles.row}>
                                                    <Text style={styles.label}>% change in 24h:</Text>
                                                    <Text style={styles.value}> {Math.round(data.quote.USD.percent_change_24h * 100) / 100}</Text>
                                                </View>
                                            </View>
                                        </View>
                                    )}
                                    <TouchableOpacity
                                        onPress={this.more}
                                        style={styles.btn}
                                    >
                                        {
                                            this.state.clicked ?
                                                <Text style={styles.btnText}>Loading...</Text>
                                                :
                                                <Text style={styles.btnText}>Show more</Text>
                                        }
                                    </TouchableOpacity>
                                </ScrollView>
                            }
                        </View>
                }
            </View>
        );
    }
}

const mapDispatchToProps = dispatch => ({
    fetchCrypto: () => dispatch(fetchCrypto()),
    fetchMoreCrypto: (start) => dispatch(fetchMoreCrypto(start)),
});

const mapStateToProps = state => ({
    data: state.home.data,
    dataMore: state.home.dataMore,
    error: state.home.error
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Home);

const styles = StyleSheet.create({
    header: {
        backgroundColor: '#1B2845',
        justifyContent: 'space-around',
        borderBottomColor: '#1B2845',
        elevation: 10
    },
    body: {
        paddingTop: 25,
        paddingRight: 25,
        paddingLeft: 25,
        paddingBottom: 180
    },
    title: {
        color: '#fff',
        fontSize: 24,
        textAlign: 'center',
        marginBottom: 10
    },
    divider: {
        borderWidth: 0.4,
        borderBottomColor: '#304D6D',
        marginBottom: 10
    },
    card: {
        borderRadius: 8,
        padding: 15,
        marginBottom: 25,
        backgroundColor: '#1B2845',
        overflow: 'hidden'
    },
    cardTop: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between'
    },
    cardBody: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginTop: 20,
    },
    row: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center'
    },
    rowSingle: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        marginTop: 20
    },
    label: {
        color: '#A7CCED',
        fontSize: 12,
    },
    value: {
        color: '#fff',
        fontSize: 16,
    },
    btn: {
        backgroundColor: '#1B2845',
        borderRadius: 8,
        justifyContent: 'center',
        paddingTop: 10,
        paddingBottom: 10,
        paddingLeft: 20,
        paddingRight: 20
    },
    btnText: {
        color: '#fff',
        fontSize: 14,
        textAlign: "center"
    },
    error: {
        justifyContent: 'center',
        height: '93%',
        padding: 25
    },
    errorText: {
        alignSelf: 'center',
        color: '#fff',
        marginBottom: 30
    }
});
