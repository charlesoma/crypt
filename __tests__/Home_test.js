import React from 'react';
import renderer from 'react-test-renderer';
import { Home } from '../components/Home';

jest.useFakeTimers();

describe('<Home />', () => {
    it('renders without crashing', () => {
        const rendered = renderer.create(<Home />).toJSON();
        expect(rendered).toBeTruthy();
    });
    test('renders correctly', () => {
        const tree = renderer.create(<Home />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
