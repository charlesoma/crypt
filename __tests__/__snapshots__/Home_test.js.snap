// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<View>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1B2845",
        "borderBottomColor": "#1B2845",
        "borderBottomWidth": 0.5,
        "elevation": 10,
        "flexDirection": "row",
        "height": 64,
        "justifyContent": "space-around",
        "paddingHorizontal": 10,
        "paddingTop": 20,
      }
    }
  >
    <Image
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 64,
            "width": undefined,
          },
          undefined,
        ]
      }
      testID="headerContainer"
    />
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 3,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
        text="Crypt"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        Crypt
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
        }
      }
    />
  </View>
  <View>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "paddingBottom": 180,
          "paddingLeft": 25,
          "paddingRight": 25,
          "paddingTop": 25,
        }
      }
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
    >
      <RCTRefreshControl />
      <View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#1B2845",
              "borderRadius": 8,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 10,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 14,
                "textAlign": "center",
              }
            }
          >
            Show more
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
